pyo3-ffi/src/object.rs

```diff
@@ -394,37 +441,161 @@ pub const Py_TPFLAGS_DICT_SUBCLASS: c_ulong = 1 << 29;
 pub const Py_TPFLAGS_BASE_EXC_SUBCLASS: c_ulong = 1 << 30;
 pub const Py_TPFLAGS_TYPE_SUBCLASS: c_ulong = 1 << 31;
 
-pub const Py_TPFLAGS_DEFAULT: c_ulong =
-    Py_TPFLAGS_HAVE_STACKLESS_EXTENSION | Py_TPFLAGS_HAVE_VERSION_TAG;
+pub const Py_TPFLAGS_DEFAULT: c_ulong = if cfg!(Py_3_10) {
+    Py_TPFLAGS_HAVE_STACKLESS_EXTENSION
+} else {
+    Py_TPFLAGS_HAVE_STACKLESS_EXTENSION | Py_TPFLAGS_HAVE_VERSION_TAG
+};
 
 pub const Py_TPFLAGS_HAVE_FINALIZE: c_ulong = 1;
+pub const Py_TPFLAGS_HAVE_VERSION_TAG: c_ulong = 1 << 18;
 
-// skipped _Py_RefTotal
-// skipped _Py_NegativeRefCount
+#[cfg(all(py_sys_config = "Py_REF_DEBUG", not(Py_LIMITED_API)))]
+extern "C" {
+    pub fn _Py_NegativeRefCount(filename: *const c_char, lineno: c_int, op: *mut PyObject);
+    #[cfg(Py_3_12)]
+    #[link_name = "_Py_IncRefTotal_DO_NOT_USE_THIS"]
+    fn _Py_INC_REFTOTAL();
+    #[cfg(Py_3_12)]
+    #[link_name = "_Py_DecRefTotal_DO_NOT_USE_THIS"]
+    fn _Py_DEC_REFTOTAL();
+}
 
 extern "C" {
     #[cfg_attr(PyPy, link_name = "_PyPy_Dealloc")]
     pub fn _Py_Dealloc(arg1: *mut PyObject);
+
+    #[cfg_attr(PyPy, link_name = "PyPy_IncRef")]
+    pub fn Py_IncRef(o: *mut PyObject);
+    #[cfg_attr(PyPy, link_name = "PyPy_DecRef")]
+    pub fn Py_DecRef(o: *mut PyObject);
+
+    #[cfg(Py_3_10)]
+    #[cfg_attr(PyPy, link_name = "_PyPy_IncRef")]
+    pub fn _Py_IncRef(o: *mut PyObject);
+    #[cfg(Py_3_10)]
+    #[cfg_attr(PyPy, link_name = "_PyPy_DecRef")]
+    pub fn _Py_DecRef(o: *mut PyObject);
 }
 
-// Reference counting macros.
-#[inline]
+#[inline(always)]
 pub unsafe fn Py_INCREF(op: *mut PyObject) {
-    if cfg!(py_sys_config = "Py_REF_DEBUG") {
-        Py_IncRef(op)
-    } else {
-        (*op).ob_refcnt += 1
+    #[cfg(any(
+        all(Py_LIMITED_API, Py_3_12),
+        all(
+            py_sys_config = "Py_REF_DEBUG",
+            Py_3_10,
+            not(all(Py_3_12, not(Py_LIMITED_API)))
+        )
+    ))]
+    {
+        return _Py_IncRef(op);
+    }
+
+    #[cfg(all(py_sys_config = "Py_REF_DEBUG", not(Py_3_10)))]
+    {
+        return Py_IncRef(op);
+    }
+
+    #[cfg(any(
+        not(Py_LIMITED_API),
+        all(Py_LIMITED_API, not(Py_3_12)),
+        all(py_sys_config = "Py_REF_DEBUG", Py_3_12, not(Py_LIMITED_API))
+    ))]
+    {
+        #[cfg(all(Py_3_12, not(target_pointer_width = "32")))]
+        {
+            let cur_refcnt = (*op).ob_refcnt.ob_refcnt_split[crate::PY_BIG_ENDIAN];
+            let new_refcnt = cur_refcnt.wrapping_add(1);
+            if new_refcnt == 0 {
+                return;
+            }
+            (*op).ob_refcnt.ob_refcnt_split[crate::PY_BIG_ENDIAN] = new_refcnt;
+        }
+
+        #[cfg(all(Py_3_12, target_pointer_width = "32"))]
+        {
+            if _Py_IsImmortal(op) != 0 {
```

Comment:
Couldn't this part be shared as in the decrement implementation?

Reply:
In the header it's split like this where the "immortality" in the 64-bit implementation is achieved by the `wrapping_add` to the lower 32 bits will hit 0, and checking for it.

https://github.com/python/cpython/blob/e3b5ed7b1cb8ad12aa285cf78303c6dda0953181/Include/object.h#L640-L654

I agree it looks a bit funky, presumably there was a technical reason this was preferred. I'd like to match the C implementation as faithfully as possible.

(I'm also hoping that when nogil comes along the whole macro is removed so that we don't have to try to reproduce it here 😂)